src/LibraBFT/Impl/Consensus/BlockStorage/Properties/BlockTree.agda

```diff
@@ -259,3 +253,94 @@ module insertQuorumCertESpec
         proj₂ contract-step₃' _ = ContractOk-trans
                                     (mkContractOk (∈Post⇒∈PreOr'-refl _∈BlockTree_ _))
                                     (contract-cont1' bt0 [])
+
+module insertQuorumCertE-ASTSpec
+  (qc : QuorumCert) (bt0  : BlockTree) where
+  open insertQuorumCertE-AST qc bt0
+
+  -- The following definitions are the same as for the EitherD version above
+  record ContractOk (btPre btPost : BlockTree) (ilPre ilPost : List InfoLog) : Set where
+    constructor mkContractOk
+    field
+      noNewQCs : ∈Post⇒∈PreOrBT (_≡ qc) btPre btPost
+
+  ContractOk-trans : ∀ {btPre btInt btPost ilPre ilInt ilPost}
+                   → ContractOk btPre btInt  ilPre ilInt
+                   → ContractOk btInt btPost ilInt ilPost
+                   → ContractOk btPre btPost ilPre ilPost
+  ContractOk-trans (mkContractOk noNewQCs) (mkContractOk noNewQCs₁) =
+                    mkContractOk (∈Post⇒∈PreOr'-trans _∈BlockTree_ (_≡ qc) noNewQCs noNewQCs₁)
+
+  Contract : Either ErrLog (BlockTree × List InfoLog) → Set
+  Contract (Left x) = ⊤
+  Contract (Right (bt1 , il)) = ContractOk bt0 bt1 [] il
+
+  -- Here we prove the weakest precondition for insertQuorumCertE-AST to ensure Contract The proof
+  -- is similar to the corresponding one for insertQuorumCertE above, but the AST framework makes it
+  -- easier to discover the proof obligations.  Note that the code for insertQuorumCertE-AST is not
+  -- broken explicitly into steps because we don't need to write the types of the individual proofs
+  -- because the framework presents them to us.
+  contract-AST : ASTPredTrans.predTrans EitherPTExt insertQuorumCertE-AST Contract unit
+  contract-AST with safetyInvariant
+  --  The Left/bail cases are easy, as Contract simply requires ⊤ in this case.
+  ... | Left _ = tt
+  ... | Right unit =
+      -- Two proof obligations are introduced due to the use of maybeSD; the AST framework with the
+      -- branching extensions ensures that the two goals are created by typing C-c C-r in a hole
+      -- here.  To understand why, see the definition of maybeSD, which is in terms of the maybeD
+      -- field of the relevant MonadMaybeD instance (EitherDASTExt-MonadMaybeD), which translates to
+      -- ASTop (Right (BCmaybe mb)).  The opPT field of BranchPT establishes that we require two
+      -- proofs, one for the nothing case and one for the just case; each receives evidence that mb
+      -- is the relevant case (nothing or just something), and determines the relevant proof
+      -- obligation.
+      const tt , λ block _ →
```

Comment:
* Added blank lines.
* Yes, correct, we do not need to pattern match on (or name) the provided evidence.  It could equally be `(λ _ → tt)` (parentheses required) but I think `const tt` is marginally clearer and saves keystrokes :-).  Plus it eliminates two unicode characters, surely this fills your heart with joy? :smile:
* I don't like calling an `ExecutedBlock` `block` either, but this is faithful to the Haskell code (and perhaps the original Rust code?), so I'm not changing it here (and in a number of other places where the same is true).